apps/docs/refactor.md

```diff
@@ -0,0 +1,629 @@
+- do not reference DOM nodes by ID or css selectors except for those DOM nodes
+  that were created and rendered to the DOM by the library.
+- do not depend directly on other feature libraries unless you are reponsible
+  for configuring them
+
+## Further Reading ##
+
+If all this sounds a bit confusing to you, read
+http://jaysoo.ca/2016/02/28/organizing-redux-application/, which is a more
+concise explanation of basically the same thing as is proposed in this document.
+
+# The Actual Plan #
+
+While having patterns for capsule and feature libraries is helpful, we need to
+go one step further to figure out how to actually execute this refactoring.
+
+## Directory Hierarchy ##
+
+The first step is determining a sensible directory structure that organizes code
+based on the various patterns we want to use, and based on the various problem
+domains that our code base covers.
+
+Here is a strawman to discuss:
+
+```
+src/
+  sites/   <--- resusable libraries do not belong in here
+    <site-name>/
+      pages/
+        <page-name>.js
+        <page-name-with-multiple-entry-points>/
+          <entry-point-name>.js
+          <other-entry-point-name>.js
+      init/
+        redux.js
+        constants.js
+        other-site-specific-config-shared-across-entry-points.js
+  lib/   <--- all reusable libraries go in here
+    kits/   <--- Big feature libraries that are only referenced from entry points
+      <kit-name>/
+        <kit-name>.js
+        redux.js
+        ui/  <--- capsules that generate DOM specific to this kit
+        util/  <--- capsules specific to this kit
+    tools/   <--- Small feature libraries that can be used by other feature libraries
+      <tool-name>/
+        <tool-name>.js
+        redux.js
+        ui/
+        util/
+    ui/   <--- capsule libraries that generate DOM
+      <ReactComponentName>.jsx
+    util/  <--- capsule libraries that do not generate DOM
+      <stateless-util-library>.js
+    third-party/
+```
+
+And here is a list of ordered steps to take that will move us in this direction:
+
+1. Move all entry points into `sites` directory structure [easy]
+2. Move initialization code into `sites` directory structure [medium]
+3. Move "leaf nodes" of depdency graph into `lib/util` directory [easy]
+4. Rename templates directory to `lib/ui` [easy]
+5. Remove cycles from our dependency graph [hard]
+6. Move applab/gamelab/weblab/maze/studio/etc into `lib/kits` directory [medium]
+7. Move all the other stateless utility modules into `lib/util` [easy]
+8. Move all the other stateful modules into `lib/tools` directory [medium]
+9. Refactor entry points to inject dependent modules into feature libraries [hard]
+10. [...]
+11. profit!
+
+The [easy] stuff should take no more than one developer day to do.
+
+The [medium] stuff should take a few developer days to do.
+
+The [hard] stuff will probably take at least a developer week to do.
+
+The [...] stuff could take an immeasurable amount of time to do.
```

Comment:
Looks like 4-5 developer-weeks of time, total.  Maximum rate of change is probably capped by how disruptive all the moving around may be.  Maybe with three of us working on it we could do this in a couple of weeks?